compiler/noirc_frontend/src/parser/parser/expression.rs

```diff
@@ -0,0 +1,1708 @@
+        };
+        assert_eq!(if_expr.condition.to_string(), "true");
+        assert!(if_expr.alternative.is_some());
+    }
+
+    #[test]
+    fn parses_parses_if_else_if() {
+        let src = "if true { 1 } else if false { 2 } else { 3 }";
+        let mut parser = Parser::for_str(src);
+        let expr = parser.parse_expression_or_error();
+        assert_eq!(expr.span.end() as usize, src.len());
+        expect_no_errors(&parser.errors);
+        let ExpressionKind::If(if_expr) = expr.kind else {
+            panic!("Expected if");
+        };
+        assert_eq!(if_expr.condition.to_string(), "true");
+        let ExpressionKind::If(..) = if_expr.alternative.unwrap().kind else {
+            panic!("Expected if");
+        };
+    }
+
+    #[test]
+    fn parses_cast() {
+        let src = "1 as u8";
+        let mut parser = Parser::for_str(src);
+        let expr = parser.parse_expression_or_error();
+        assert_eq!(expr.span.end() as usize, src.len());
+        expect_no_errors(&parser.errors);
+        let ExpressionKind::Cast(cast_expr) = expr.kind else {
+            panic!("Expected cast");
+        };
+        assert_eq!(cast_expr.lhs.to_string(), "1");
+        assert_eq!(cast_expr.r#type.to_string(), "u8");
+    }
+
+    #[test]
+    fn parses_cast_missing_type() {
+        let src = "
+        1 as 
+            ^
+        ";
+        let (src, span) = get_source_with_error_span(src);
+        let mut parser = Parser::for_str(&src);
+        parser.parse_expression();
+        let error = get_single_error(&parser.errors, span);
+        assert_eq!(error.to_string(), "Expected a type but found end of input");
+    }
+
+    #[test]
+    fn parses_index() {
+        let src = "1[2]";
+        let mut parser = Parser::for_str(src);
+        let expr = parser.parse_expression_or_error();
+        assert_eq!(expr.span.end() as usize, src.len());
+        expect_no_errors(&parser.errors);
+        let ExpressionKind::Index(index_expr) = expr.kind else {
+            panic!("Expected index");
+        };
+        assert_eq!(index_expr.collection.to_string(), "1");
+        assert_eq!(index_expr.index.to_string(), "2");
+    }
+
+    #[test]
+    fn parses_operators() {
+        for operator in BinaryOpKind::iter() {
+            let src = format!("1 {operator} 2");
+            let mut parser = Parser::for_str(&src);
+            let expr = parser.parse_expression_or_error();
+            assert_eq!(expr.span.end() as usize, src.len());
+            assert!(parser.errors.is_empty(), "Expected no errors for {operator}");
+            let ExpressionKind::Infix(infix_expr) = expr.kind else {
+                panic!("Expected infix for {operator}");
+            };
+            assert_eq!(infix_expr.lhs.to_string(), "1");
+            assert_eq!(infix_expr.operator.contents, operator);
+            assert_eq!(infix_expr.rhs.to_string(), "2");
+        }
+    }
+
+    #[test]
+    fn parses_operator_precedence() {
```

Comment:
This is the only test that (directly) checks how the parser handles operator precedence that we have currently. Can you improve coverage here to ensure that we handle mixes of arithmetic, bitwise, equality, bitshift, modulo operators correctly?

Reply:
Sounds good! Do you know if there's an existing test for the chumsky parser?

Reply:
Nevermind, I think I know how to make a simple and good test for this.

Reply:
Done! Let me know if the test I added is good.